File: debop4s-timeperiod/src/test/scala/com/github/debop4s/timeperiod/tests/timeranges/WeekRangeCollectionTest.scala
package com.github.debop4s.timeperiod.tests.timeranges


import com.github.debop4s.timeperiod._
import com.github.debop4s.timeperiod.tests.AbstractTimePeriodTest
import com.github.debop4s.timeperiod.timerange.{WeekRange, WeekRangeCollection}
import com.github.debop4s.timeperiod.utils.{Weeks, Times}

/**
 * WeekRangeCollectionTest
 * Created by debop on 2014. 2. 16.
 */
class WeekRangeCollectionTest extends AbstractTimePeriodTest {

    test("single months") {
        val startYear = 2004
        val startWeek = 22

        val wrs = new WeekRangeCollection(startYear, startWeek, 1)
        wrs.weekCount should equal(1)

        wrs.startYear should equal(startYear)
        wrs.endYear should equal(startYear)
        wrs.startWeekOfYear should equal(startWeek)
        wrs.endWeekOfYear should equal(startWeek)

        val weeks = wrs.getWeeks
        weeks should have size 1
        weeks(0).isSamePeriod(WeekRange(startYear, startWeek)) should equal(true)
    }

    test("calendar weeks") {
        val startYear = 2004
        val startWeek = 22
        val weekCount = 5

        val wrs = WeekRangeCollection(startYear, startWeek, weekCount)

        wrs.weekCount should equal(weekCount)
        wrs.startYear should equal(startYear)
        wrs.startWeekOfYear should equal(startWeek)
        wrs.endYear should equal(startYear)
        wrs.endWeekOfYear should equal(startWeek + weekCount - 1)
    }

    test("weekCounts") {
        val weekCounts = Array(1, 6, 48, 180, 360)

        val now = Times.now
        val today = Times.today

        weekCounts.foreach(weekCount => {
            val wrs = WeekRangeCollection(now, weekCount)

            val startTime = wrs.calendar.mapStart(Times.startTimeOfWeek(today))
            val endTime = wrs.calendar.mapEnd(startTime + weekCount.week)

            wrs.start should equal(startTime)
            wrs.end should equal(endTime)

            val items = wrs.getWeeks

            (0 until weekCount).par.foreach(w => {
                val item = items(w)

                item.start should equal(startTime.plusWeeks(w))
                item.end should equal(wrs.calendar.mapEnd(startTime + (w + 1).week))

                item.unmappedStart should equal(startTime + w.week)
                item.unmappedEnd should equal(startTime + (w + 1).week)

                item.isSamePeriod(WeekRange(wrs.start + w.week)) should equal(true)

                val yw = Weeks.addWeekOfYears(now.getWeekyear, now.getWeekOfWeekyear, w)
                println(s"item=$item, yw=${WeekRange(yw)}")
                item.isSamePeriod(WeekRange(yw)) should equal(true)
            })
        })
    }
}

File: debop4s-timeperiod/src/test/java/debop4s/timeperiod/tests/timeranges/WeekRangeCollectionTest.java
package debop4s.timeperiod.tests.timeranges;

import debop4s.core.JAction1;
import debop4s.core.parallels.JParallels;
import debop4s.timeperiod.YearWeek;
import debop4s.timeperiod.tests.TimePeriodTestBase;
import debop4s.timeperiod.timerange.WeekRange;
import debop4s.timeperiod.timerange.WeekRangeCollection;
import debop4s.timeperiod.utils.Times;
import debop4s.timeperiod.utils.Weeks;
import lombok.extern.slf4j.Slf4j;
import org.joda.time.DateTime;
import org.junit.Test;

import java.util.List;

import static org.fest.assertions.Assertions.assertThat;

/**
 * kr.hconnect.timeperiod.test.timeranges.WeekRangeCollectionFunSuite
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 13. 5. 25. 오후 11:49
 */
@Slf4j
public class WeekRangeCollectionTest extends TimePeriodTestBase {

    @Test
    public void singleMonths() {
        final int startYear = 2004;
        final int startWeek = 22;

        WeekRangeCollection weekRanges = new WeekRangeCollection(startYear, startWeek, 1);
        assertThat(weekRanges.getWeekCount()).isEqualTo(1);

        assertThat(weekRanges.getStartYear()).isEqualTo(startYear);
        assertThat(weekRanges.getEndYear()).isEqualTo(startYear);
        assertThat(weekRanges.getStartWeekOfYear()).isEqualTo(startWeek);
        assertThat(weekRanges.getEndWeekOfYear()).isEqualTo(startWeek);

        List<WeekRange> weeks = weekRanges.getWeeks();
        assertThat(weeks.size()).isEqualTo(1);
        assertThat(weeks.get(0).isSamePeriod(new WeekRange(startYear, startWeek))).isTrue();
    }

    @Test
    public void calenarWeeks() {
        final int startYear = 2004;
        final int startWeek = 22;
        final int weekCount = 5;

        WeekRangeCollection weekRanges = new WeekRangeCollection(startYear, startWeek, weekCount);

        assertThat(weekRanges.getWeekCount()).isEqualTo(weekCount);
        assertThat(weekRanges.getStartYear()).isEqualTo(startYear);
        assertThat(weekRanges.getStartWeekOfYear()).isEqualTo(startWeek);
        assertThat(weekRanges.getEndYear()).isEqualTo(startYear);
        assertThat(weekRanges.getEndWeekOfYear()).isEqualTo((startWeek + weekCount - 1));
    }

    @Test
    public void weeksCountsTest() {
        int[] weekCounts = new int[] { 1, 6, 48, 180, 365 };

        final DateTime now = Times.now();
        final DateTime today = Times.today();

        for (int weekCount : weekCounts) {
            final WeekRangeCollection weekRanges = new WeekRangeCollection(now, weekCount);

            final DateTime startTime = weekRanges.getTimeCalendar().mapStart(Times.startTimeOfWeek(today));
            final DateTime endTime = weekRanges.getTimeCalendar().mapEnd(startTime.plusWeeks(weekCount));

            assertThat(weekRanges.getStart()).isEqualTo(startTime);
            assertThat(weekRanges.getEnd()).isEqualTo(endTime);

            final List<WeekRange> items = weekRanges.getWeeks();

            JParallels.run(weekCount, new JAction1<Integer>() {
                @Override
                public void perform(Integer w) {
                    final WeekRange item = items.get(w);
                    assertThat(item.getStart()).isEqualTo(startTime.plusWeeks(w));
                    assertThat(item.getEnd()).isEqualTo(weekRanges.getTimeCalendar().mapEnd(startTime.plusWeeks(w + 1)));

                    assertThat(item.getUnmappedStart()).isEqualTo(startTime.plusWeeks(w));
                    assertThat(item.getUnmappedEnd()).isEqualTo(startTime.plusWeeks(w + 1));

                    assertThat(item.isSamePeriod(new WeekRange(weekRanges.getStart().plusWeeks(w)))).isTrue();

                    YearWeek yw = Weeks.addWeekOfYears(now.getWeekyear(), now.getWeekOfWeekyear(), w);
                    assertThat(item.isSamePeriod(new WeekRange(yw.weekyear(), yw.weekOfWeekyear()))).isTrue();
                }
            });
        }
    }
}
